src/participant/prompts/promptHistory.ts

```diff
@@ -0,0 +1,198 @@
+      nextTurn.result.errorDetails?.message === ParticipantErrorTypes.FILTERED
+    ) {
+      // If the response to this request led to a filtered error,
+      // we do not want to include it in the history
+      return undefined;
+    }
+
+    if (
+      currentTurn.prompt?.trim().length === 0 ||
+      connectionNames?.includes(currentTurn.prompt)
+    ) {
+      // When the message is empty or a connection name then we skip it.
+      // It's probably going to be the response to the connect step.
+      return undefined;
+    }
+
+    // eslint-disable-next-line new-cap
+    return vscode.LanguageModelChatMessage.User(currentTurn.prompt);
+  }
+
+  /** When passing the history to the model we only want contextual messages
+  to be passed. This function parses through the history and returns
+  the messages that are valuable to keep. */
+  static getFilteredHistory({
+    connectionNames,
+    history,
+    databaseName,
+    collectionName,
+  }: {
+    connectionNames?: string[]; // Used to scrape the connecting messages from the history.
+    history?: vscode.ChatContext['history'];
+    databaseName?: string;
+    collectionName?: string;
+  }): vscode.LanguageModelChatMessage[] {
+    const messages: vscode.LanguageModelChatMessage[] = [];
+
+    if (!history) {
+      return [];
+    }
+
+    const namespaceIsKnown =
+      databaseName !== undefined && collectionName !== undefined;
+    for (let i = 0; i < history.length; i++) {
+      const currentTurn = history[i];
+
+      let addedMessage: vscode.LanguageModelChatMessage | undefined;
+      if (currentTurn instanceof vscode.ChatRequestTurn) {
+        const previousTurn = i - 1 >= 0 ? history[i - 1] : undefined;
+        const nextTurn = i + 1 < history.length ? history[i + 1] : undefined;
+
+        addedMessage = this._handleChatRequestTurn({
+          previousTurn,
+          currentTurn,
+          nextTurn,
+          connectionNames,
+          namespaceIsKnown,
+        });
+      } else if (currentTurn instanceof vscode.ChatResponseTurn) {
+        addedMessage = this._handleChatResponseTurn({
+          currentTurn,
+          namespaceIsKnown,
+        });
+      }
+      if (addedMessage) {
+        messages.push(addedMessage);
+      }
+    }
+
+    return messages;
+  }
+
+  /** The docs chatbot keeps its own history so we avoid any
+   * we need to include history only since last docs message. */
+  static getFilteredHistoryForDocs({
+    connectionNames,
+    context,
+    databaseName,
+    collectionName,
+  }: {
+    connectionNames?: string[];
+    context?: vscode.ChatContext;
+    databaseName?: string;
+    collectionName?: string;
+  }): vscode.LanguageModelChatMessage[] {
+    if (!context) {
+      return [];
+    }
+    const historySinceLastDocs: (
+      | vscode.ChatRequestTurn
+      | vscode.ChatResponseTurn
+    )[] = [];
+
+    /** Limit included messages' history to prevent prompt overflow. */
+    const MAX_DOCS_HISTORY_LENGTH = 4;
```

Comment:
You were thinking about 2-4 docs. Wondering, did 4 work better? Like a bigger content bring more relevant results?

Reply:
To be honest, I did not see a big difference, it was more of a question of at what point would it be too much content for the prompt, on average. The only real usefulness I can think of for history with docs is when someone runs i.e. queries and then says something like `/docs what does this do?` i.e. providing context to "this" is useful.

In that sense, 4 seemed fine as it's possible you could do something like `/query find all documents with age > 10` and then run `/query make it also filter for name` and then maybe do something like`/docs tell me more about this` then it _could_ be useful to have context of first message. Of course similarly there might be 3+ 

I think either way, it'd be quite fine.